### DISCUSSION TITLE
KOReader performance vs Pocketbook reader

### DISCUSSION THREAD

**Discussion by dmalinovsky** [koreader/koreader#12897] [General]
Recently I had an idea to measure KOReader page turn speed on my Pocketbook Inkpad Color 3 and compare it with the built-in software. I borrowed poor man's comparison methodology [here](https://www.mobileread.com/forums/showpost.php?p=4145822&postcount=8), using a simple ePub file without images. Full screen refresh was turned OFF. I was using taps to turn the pages, since it's easier to detect on the camera, double taps are turned off.

In my amateur tests, KOReader performed similar to the stock software, turning one page in about 2/3 sec, which is very nice. As I understand, KOReader is using single process, thus loading only one CPU core, while Pocketbook software must be using all 4 cores, because its battery consumption is way higher. I didn't measure the exact difference yet, but according to its battery stats, KOReader is consuming about 1% of battery on average (and this is with the bright enough frontlight to compensate for Kaleido 3 screen darkness), while PBReader eats several times more.

We also have latest Kindle Colorsoft in the family, so out of curiosity I measured its page turn speed as well — 1/3 sec. I'm not sure how Amazon achieved these results, taking into account they're using 2 CPU cores vs 4 in Pocketbook (see [comparison](https://comparisontabl.es/e-readers/?product1=PocketBook+InkPad+Color+3&product2=Kindle+Colorsoft+Signature+Edition#sbs)). I wonder if anybody has insights on this.

Overall, kudos to the KOReader developers for the great work! As a typical OCD guy, I appreciate this app's configurability and humble power use. 🎉 👍 

P.S. Kindle Oasis 3 takes 1/2 sec to turn the page, but its CPU is [twice slower](https://comparisontabl.es/e-readers/?product1=Kindle+Oasis+3&product2=Kindle+Colorsoft+Signature+Edition#sbs) than Colorsoft's one.

**Comment by dmalinovsky** [comment]
The official Kindle Colorsoft press release [said](https://www.aboutamazon.com/news/devices/kindle-color-specs-price-announce):

> We built the display on an oxide backplane for sharper contrast, faster page turns, and better image quality.

Looks like it's one of the reasons for the better performance.

**Comment by NiLuJe** [comment]
PB is a special snowflake: we're (mostly) forced to use PB's own middleware, InkView.

This *severely* hampers our ability to do things better than the stock system in terms of UX latency.

**Reply by NiLuJe** [reply]
Which is why comparing anything else (other than Android) to a PB will always *destroy* the PB. *Especially* a Kindle, which have always been the gold standard (barring fucking up the stock UI with each refactor) ;).

**Reply by dmalinovsky** [reply]
I also measured Kindle Oasis 3 and it takes 1/2 sec to turn the page on it — so it's slower than Colorsoft but faster than Pocketbook.

> PB is a special snowflake: we're (mostly) forced to use PB's own middleware, InkView.

Yeah, I know about this. I was surprised that PBReader is relatively sluggish, they have decent hardware, it seems.

I've no complaints about KOReader. :)

**Reply by dmalinovsky** [reply]
Besides, Kindle's software is written in Java, while Pocketbook is using C++, as I understood. And the latter is supposed to be more performant language. 🤷‍♂️

**Reply by pazos** [reply]
Java is just an interpreted language, what matters is the implementation ;). 

Kindles and android might run a (very old) compatible java virtual machine but its implementation is C++ on steroids, very well optimized.

Basically they're stripping away the "write once, run everywhere" slogan. A java program won't run on android or a kindle. 

A source code with java syntax that uses vendor provided classes might work if built with proper sdks.

We must remind that Lab126 employed genious like the [guy who runs linux on an intel 4004](https://dmitry.gr/?r=05.Projects&proj=35.%20Linux4004)

**Reply by dmalinovsky** [reply]
One of the things I don't like in Pocketbook and its built-in reader is the touch input latency: whenever you need to do any interaction with the touch screen, it's taking a while, except the simple taps. And my Pocketbook has the best CPU among [all non-Android ones](https://comparisontabl.es/pocketbook-e-readers/)!

Buttons work quickly enough, which I'm using most of the time.

**Comment by NiLuJe** [comment]
The UI was in Java, not the backends ;). They've since switches to React
native in the new UI, still some Java left, but much less ;).

**Reply by dmalinovsky** [reply]
Interesting, I didn't know that, now it makes more sense. Also, they don't have to support full ePub 3 specification, but neither does KOReader. :)

**Reply by Frenzie** [reply]
> React Native 

Now that's much slower than Java. ;-P

**Reply by pazos** [reply]
These days web tech is quite awesome. Have you tried flutter?

IMHO for UI stuff there's no slow tool nowadays :)

**Comment by minttea1299** [comment]
Thanks for sharing your findings. Would you care to replicate the test for word lookup with dictionary and text highlighting? 
I've seen many mentioning the sluggish response on Pocketbook devices. Curious to see if KOReader can help for this particular type of action

**Reply by dmalinovsky** [reply]
So if I turn on dictionary mode in PBReader, where you can lookup a word with a single tap, it takes about a second from tapping till showing the word definition. For KOReader it took about 1 1/4 sec, but I have 4 dictionaries enabled there, so I think the timing is similar. 

As for highlighting, PBReader didn’t allow me to select several words from the first attempt — I could only select one word and then used selection handles around the word to select more. Overall, it was a cumbersome experience. It’s much more robust in KOReader, it has no problem selecting multiple words at once.

**Reply by minttea1299** [reply]
Thank you for sharing 🙏 It's great to see KOReader can bring better experience to more affordable devices

**Comment by dmalinovsky** [comment]
So I've ordered freshly released Pocketbook Era Lite and compared refresh speeds for both programs. Here it got interesting: on the same ePub file PBReader took 3/4 sec on average, while for KOReader it was only 1/2 sec.

Way to go!

**Comment by NhzRzn** [comment]
On my poketbook era color, the standard reader flips pages 3 times faster than the koreader, the standard reader takes an instant to flip through, and the koreader flips in about 1 second, although sometimes, for no apparent reason, the koreader works as fast as the standard reader, but this rarely happens.

**Reply by dmalinovsky** [reply]
Yes, I have the same problem on my color Pocketbook Inkpad Color 3, which has the same processor as your ereader. It seems that after opening a new book, KOReader needs some "warming up". Try opening up any dialog window and close it without making any changes, e.g. Gear > Device > Auto-standby settings > Min input idle seconds. Then click Close. Make sure you do that *after* opening a book.

After that, KOReader should handle the book as fast as the built-in reader. I have some suspicions about why it may happen — too aggressive CPU power management doesn't give enough time for Lua JIT caching, etc., but couldn't find a way to solve it once and for all.

**Reply by NhzRzn** [reply]
it works 1 time out of 100

**Reply by dmalinovsky** [reply]
If you have some processor hungry user plugins enabled (like Bookends), it's better to disable them.
